crates/oxc_linter/src/lint_adapter.rs

```diff
@@ -0,0 +1,1208 @@
+        resolve_coercion_with, resolve_neighbors_with, resolve_property_with, Adapter,
+        ContextIterator, ContextOutcomeIterator, EdgeParameters, ResolveEdgeInfo, ResolveInfo,
+        TrustfallEnumVertex, VertexIterator,
+    },
+    FieldValue,
+};
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct InputQuery {
+    pub query: String,
+    pub args: BTreeMap<Arc<str>, FieldValue>,
+    pub reason: String,
+}
+
+#[derive(Debug, Clone, TrustfallEnumVertex)]
+pub enum Vertex<'a> {
+    AstNode(AstNode<'a>),
+    Expression(&'a Expression<'a>),
+    Argument(&'a Argument<'a>, i32), // i32 is argument index
+    Span(Span),
+    Parameter(&'a BindingPatternKind<'a>, i32), // i32 is argument index
+    PossiblyNumber(f64),
+    PropertyAccess(Span, &'a str),
+    PropertyKey(&'a PropertyKey<'a>),
+    TSType(&'a TSType<'a>),
+    TSSignature(&'a TSSignature<'a>),
+    TSTypeAnnotation(&'a TSTypeAnnotation<'a>),
+    TSLiteral(&'a TSLiteral<'a>, &'a Span),
+    TSTypeName(&'a TSTypeName<'a>, &'a Span),
+    EffectivelyLiteralString(&'a Span, &'a Atom),
+    ImportSpecifier(&'a ImportDeclarationSpecifier),
+    JsxElementName(&'a JSXElementName<'a>),
+    JsxOpeningElement(&'a JSXOpeningElement<'a>),
+    JsxClosingElement(&'a JSXClosingElement<'a>),
+    ObjectProperty(&'a ObjectPropertyKind<'a>),
+    VariableDeclaration(&'a VariableDeclarator<'a>),
+    // operators
+    Operator(&'a BinaryOperator),
+    AssignmentOperator(&'a AssignmentExpression<'a>),
+    // assignment to ...
+    AssignmentToIdentifier(&'a Span, &'a Atom),
+    AssignmentOnObject(&'a SimpleAssignmentTarget<'a>),
+    DestructuringAssignment(DestructuringAssignment<'a>),
+
+    File(Rc<AstNodeIterator<'a>>),
+    // Comment(u32, u32, &'a str), // span_start, span_end, comment
+}
+
+#[derive(Debug, Clone)]
+pub enum DestructuringAssignment<'a> {
+    AssignmentExpression(&'a AssignmentTargetPattern<'a>),
+    ArrayPattern(&'a ArrayPattern<'a>),
+    ObjectPattern(&'a ObjectPattern<'a>),
+    DefaultedDestructuringAssignment(&'a AssignmentPattern<'a>),
+}
+
+impl<'a> GetSpan for DestructuringAssignment<'a> {
+    fn span(&self) -> Span {
+        match self {
+            DestructuringAssignment::AssignmentExpression(expr) => expr.span(),
+            DestructuringAssignment::ArrayPattern(expr) => expr.span,
+            DestructuringAssignment::ObjectPattern(expr) => expr.span,
+            DestructuringAssignment::DefaultedDestructuringAssignment(expr) => expr.span,
+        }
+    }
+}
+
+pub struct AstNodeIterator<'a> {
+    iter: RefCell<Option<VertexIterator<'a, Vertex<'a>>>>,
+}
+
+impl std::fmt::Debug for AstNodeIterator<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("AstNodeIterator").finish()
+    }
+}
+
+impl<'a> Clone for AstNodeIterator<'a> {
+    fn clone(&self) -> Self {
+        println!("cloning AstNodeIterator");
+        Self { iter: None.into() }
+    }
+}
+
+pub struct LintAdapter<'a> {
```

Comment:
can this `#[derive(Debug)]`?

```suggestion
#[derive(Debug)]
pub struct LintAdapter<'a> {
```

Reply:
No, Semantic doesn't implement Debug

Reply:
If I derive Debug for Semantic then yes, although thats a public api change, so I won't for now.